litmus/libdir/_aarch64/kvm_fault_type.c

```diff
@@ -1,34 +1,172 @@
+#define MAX_FAULTS_PER_THREAD 8
 
 enum fault_type_t {
-  FaultMMUAddressSize = 0,
+  FaultUndefinedInstruction,
+  FaultMMUAddressSize,
   FaultMMUTranslation,
   FaultMMUAccessFlag,
   FaultMMUPermission,
   FaultTagCheck,
   FaultUnsupported,
-  NoFault,
+  FaultUnknown,
   FaultTypes,
 };
 
 static const char *fault_type_names[] = {
+  "UndefinedInstruction",
   "MMU:AddressSize",
   "MMU:Translation",
   "MMU:AccessFlag",
   "MMU:Permission",
   "TagCheck",
   "Unsupported",
-  "NoFault"
 };
 
-static enum fault_type_t get_fault_type(unsigned int esr)
+static enum fault_type_t get_fault_type(unsigned long esr)
 {
-  unsigned int dfsc = esr & 0x3fU;
-  int fault_type = dfsc >> 2;
-  if (esr == -1)
-    return NoFault;
+  unsigned int ec;
+  unsigned int dfsc;
+  int fault_type;
 
-  if (fault_type >= FaultUnsupported)
-    return FaultUnsupported;
-  else
-    return fault_type;
+  ec = esr >> ESR_EL1_EC_SHIFT;
+  if (ec == ESR_EL1_EC_UNKNOWN) {
+    return FaultUndefinedInstruction;
+  } else {
+    dfsc = esr & 0x3fU;
+    fault_type = (dfsc >> 2) + FaultMMUAddressSize;
+    if (fault_type > FaultTagCheck)
+      return FaultUnsupported;
+    else
+      return fault_type;
+  }
+}
+
+typedef struct {
+  int instr_symb;
+  int data_symb;
+  enum fault_type_t type;
+} fault_info_t;
+
+typedef struct {
+  fault_info_t faults[MAX_FAULTS_PER_THREAD];
+  int n;
+} th_faults_info_t;
+
+static void th_faults_info_init(th_faults_info_t *th_flts)
+{
+  for (int i = 0; i < MAX_FAULTS_PER_THREAD; i++) {
+    fault_info_t *f = &th_flts->faults[i];
+    f->instr_symb = INSTR_SYMB_ID_UNKNOWN;
+    f->data_symb = DATA_SYMB_ID_UNKNOWN;
+    f->type = FaultUnknown;
+  }
+  th_flts->n = 0;
+}
+
+static int th_faults_info_compare(th_faults_info_t *th_flts1, th_faults_info_t *th_flts2)
+{
+  if (th_flts1->n != th_flts2->n)
+    return 0;
+
+  int i1 = 0, i2 = 0;
+  while (i1 < th_flts1->n) {
+    fault_info_t *f1 = &th_flts1->faults[i1];
+    fault_info_t *f2 = &th_flts2->faults[i2];
+    if (i2 == th_flts2->n)
+      return 0;
+
+    if (f1->instr_symb == f2->instr_symb && f1->data_symb == f2->data_symb &&
+        f1->type == f2->type) {
+      i1++;
+      i2 = 0;
+    } else {
+      i2++;
+    }
+  }
+  return 1;
+}
+
+static void pp_fault(int proc, int instr_symb, int data_symb, int ftype)
+{
+  printf("fault(P%d", proc);
+  if (instr_symb != INSTR_SYMB_ID_UNKNOWN)
+    printf(":%s", instr_symb_name[instr_symb]);
+  if (data_symb != DATA_SYMB_ID_UNKNOWN)
+    printf(",%s", data_symb_name[data_symb]);
+  if (ftype != FaultUnknown)
+    printf(",%s", fault_type_names[ftype]);
+  printf(");");
+}
+
+static bool fault_reported[NTHREADS][MAX_FAULTS_PER_THREAD];
+
+void pp_log_faults_init()
```

Comment:
This function must be declared static, as there may be several of them with option `-driver C`
```suggestion
static void pp_log_faults_init()
```
Then the function has to be called at least once, so as to avoid a gcc warning.